c10/core/TensorImpl.cpp

```diff
@@ -59,16 +59,18 @@ void TensorImpl::_set_fw_grad(const at::Tensor& new_grad, const at::Tensor& self
 TensorImpl::TensorImpl(
     Storage&& storage,
     DispatchKeySet key_set,
-    const caffe2::TypeMeta data_type)
+    const caffe2::TypeMeta data_type,
+    bool is_view)
     // Use std::forward to suppress static analyzer false positive.
-    : TensorImpl(std::forward<Storage>(storage), key_set, data_type, storage.device()) {}
+    : TensorImpl(std::forward<Storage>(storage), key_set, data_type, storage.device(), is_view) {}
 
 TensorImpl::TensorImpl(DispatchKeySet key_set, const caffe2::TypeMeta data_type, c10::optional<c10::Device> device_opt)
     : TensorImpl({}, key_set, data_type, std::move(device_opt)) {}
 
 TensorImpl::TensorImpl(Storage&& storage, DispatchKeySet key_set, const caffe2::TypeMeta data_type,
-                       c10::optional<c10::Device> device_opt)
+                       c10::optional<c10::Device> device_opt, bool is_view)
```

Comment:
Hmm, stop me if I already commented this. But it seems like there might be a good opportunity to simplify the logic here if you force users to directly pass in the VersionCounter they're supposed to use, in much the same way we force people to pass in the Storage. (Actually, I've always wanted to store the VersionCounter directly on Storage, but because we can detach version counts from their underlying storage they can't actually be stored this way)

Reply:
@ezyang I definitely agree it's a good opportunity to pass in VersionCounter directly. This improvement is a little tricky tho since we might still want to keep "view tensor share VariableVersion" logic which currently lives in `as_view` still in `as_view` (instead of the TensorImpl constructor which happens inside inference kernel). I might prefer to do that as a followup PR so that we can get this one in first to unblock a few more AutoNonVariableTypeMode replacements. Wdyt?

Reply:
I agree it's tricky. This is definitely one of those cases where I'd personally try to stack a PR before this one doing the refactor to see if it was possible, and if it wasn't but it was feasible in combination with this PR, I'd do it together. Well, re unblocking, let's see if I can convince myself that the way it is currently being done is correct.